lib/util.py

```diff
@@ -678,79 +679,135 @@ class TxHashMismatch(ServerError):
+        ''' Returns tuple:
+
+        read_pending - new data is available that may be unknown to select(),
+            so perform a get() regardless of select().
+        write_pending - some send data is still buffered, so make sure to call
+            send_flush if writing becomes available.
+        '''
+        try:
+            # pending() only defined on SSL sockets.
+            has_pending = self.socket.pending() > 0
+        except AttributeError:
+            has_pending = False
+        return has_pending, bool(self.send_buf)
 
     def get(self):
+        ''' Attempt to read out a message, possibly saving additional messages in
+        a receive buffer.
+
+        If no message is currently available, this raises util.timeout and you
+        should retry once data becomes available to read. If connection is bad for
+        some known reason, raises .Closed; other errors will raise other exceptions.
+        '''
         while True:
-            response, self.message = parse_json(self.message)
+            response, self.recv_buf = parse_json(self.recv_buf)
             if response is not None:
                 return response
+
             try:
                 data = self.socket.recv(1024)
-            except socket.timeout:
+            except (socket.timeout, BlockingIOError, ssl.SSLWantReadError):
                 raise timeout
-            except ssl.SSLError:
-                raise timeout
-            except socket.error as err:
-                if err.errno == 60:
-                    raise timeout
-                elif err.errno in [11, 35, 10035]:
-                    self.print_error("socket errno %d (resource temporarily unavailable)"% err.errno)
-                    time.sleep(0.2)
+            except OSError as exc:
+                if exc.errno in (11, 35, 60, 10035):
+                    # some OSes might give these ways of indicating a would-block error.
                     raise timeout
-                else:
-                    self.print_error("socket error:", err)
-                    data = b''
-            except:
-                traceback.print_exc(file=sys.stderr)
-                data = b''
-
-            if not data:  # Connection closed remotely
-                return None
-            self.message += data
+                if exc.errno == 9:
+                    # EBADF. Someone called close() locally so FD is bad.
+                    raise self.Closed('closed by local')
+                raise self.Closed('closing due to {}: {}'.format(type(exc).__name__, str(exc)))
+            except ssl.SSLError as e:
+                # Note: rarely an SSLWantWriteError can happen if we renegotiate
+                # SSL and buffers are full. This is pretty annoying to handle
+                # right and we don't expect to renegotiate, so just drop
+                # connection.
+                raise self.Closed('closing due to {}: {}'.format(type(exc).__name__, str(exc)))
+
+            if not data:
```

Comment:
Is `data` always defined for every possible combination of exceptions? It seems like it might not be always.. esp. if errno is not in the set `{9, 11, 35, 60, 10035}`?

Reply:
Or maybe it is.. sorry. :) LMK.

Reply:
Yes, it should be that no exception continues the scope (all excepts have raises in them).